### ISSUE TITLE
Symlinks to directories are not listed as directories and cannot be browsed

### ISSUE THREAD

**Issue by nburrus** [aiekick/ImGuiFileDialog#88]
**Describe the bug**
Symlinks to directories are listed as "link" files, but not as directories. This means that we cannot follow symlinks to navigate the filesystem. Every file browsers I can think of would let users enter symlinks to directories as if they were directories, so unless I missed some option I think it's a bug.

I could see rendering the symlinks to directories in italic or with a different style, but still listing them as directories.

A quick fix I use locally is to check first for the directory type, and only after for the link type:

```
if (file.is_directory())
    fileType = 'd';
else if (file.is_symlink())
    fileType = 'l';
```

This works because symlinks to directories return true to is_directory().

**To Reproduce**
I'm using the Lib_Only branch, commit 5d447f557 (March 1st, 2022)
ln -s /tmp link-to-tmp
-> Create an OpenDialog, you can't enter link-to-tmp. It'll be listed as a [Link] file if there is no filter, but selecting it will try to open it as a file.

**Expected behavior**
Follow the symlink and enter the directory.

**Desktop (please complete the following information):**
 - OS: Linux, macOS, probably Windows with WSL

**Comment by aiekick** [comment]
Hello,

ha yes your are right. thanks for the bug reporting.

will be easy to fix.

**Comment by aiekick** [comment]
its fixed on my side. can you confirm the fix ?

the next question is : do i need to add an option for style customisation when its a direcotry and a link at the same time.
since for now there is a style for file, link, dir but not combined

**Comment by nburrus** [comment]
Thanks for the quick fix (and for the nice lib :)! Just tested and it works. About the customization, it's indeed pretty common to differentiate links with a specific style (italic, small arrow overlay on the icon..). 

Actually I'd argue that it might be better to remove the explicit [Link] type prefix entirely and just render files and directories that are links with a different style. In practice I think that people expect links to behave exactly like the original file or folder.

One extra issue I just noticed is that sorting by filename is also buggy with links to files. It puts them kind of randomly because the sorting comparison function includes a `if (a->fileType != b->fileType) return (a->fileType == 'd'); // directory in first` which fails to account for the link filetype. And so the final ordering depends on the order of the comparisons.

So maybe the simplest general fix would be to reduce fileType to just f or d, and add a separate flag to know whether it's a symlink and render the style accordingly, but otherwise treat them exactly like the file or folder they point to?

**Comment by aiekick** [comment]
yep thanks, but see this answer : https://stackoverflow.com/a/23484991/3904977

it seems there is no way for know without exploring if a link is a dir or a file.

for c++17 its covered by the framework but when using dirent i not see a cross platfrom way

**Comment by nburrus** [comment]
Indeed, I think a `stat` is required to figure out what is behind the link, which is what std::filesystem does I think. I've submitted a draft PR #89 to feed this discussion. It tries to solve the problem by storing both the content type and a symlink flag for each item, so the rest of the code can then check the relevant part. Basically everything except the style just checks the content type (what the links refers to), so files and directories are handled similarly regardless of whether it's a link or not.  

Of course feel free to completely discard it if you don't like it, but sometimes it's just easier to write some code to propose something concrete :)

**Comment by aiekick** [comment]
ok thanks, will check that. np :)

**Comment by nburrus** [comment]
Thanks for continuing on this! Just tested the current version and did not spot any issue so far. I had to fix the build when USE_STD_FILESYSTEM is false though, here is the commit in my repository:  https://github.com/nburrus/ImGuiFileDialog/commit/e415a19b4ec1012af30d6a2c97803b54b06c5424 . I can do a new PR if you prefer, but not sure if it's worth the overhead since it's pretty small.

**Comment by aiekick** [comment]
yes thanks, i have pushed your fix.
btw i think i need to redesign the styling feature, and for the moment its bad for links.
but i have not mush time for the moment

**Comment by aiekick** [comment]
Hello,
i have added a FileSystem interface. you can now use your own FileSystem.
check the DemoApp who show how to use this feature with bool::filesystem
